packages/network-controller/src/NetworkController.ts

```diff
@@ -1063,205 +1482,427 @@ export class NetworkController extends BaseController<
     ) {
-      customNetworkClientRegistry[networkClientId] =
-        createAutoManagedNetworkClient({
-          type: NetworkClientType.Custom,
-          chainId,
-          rpcUrl,
-          ticker,
-        });
+      throw new Error(
+        `Could not update network: Cannot update RPC endpoints in such a way that the selected network '${this.state.selectedNetworkClientId}' would be removed without a replacement. Choose a different RPC endpoint as the selected network via the \`replacementSelectedRpcEndpointIndex\` option.`,
+      );
     }
 
-    this.update((state) => {
-      state.networkConfigurations[upsertedNetworkConfigurationId] = {
-        id: upsertedNetworkConfigurationId,
-        ...sanitizedNetworkConfiguration,
-      };
+    this.#registerNetworkClientsAsNeeded({
+      networkFields: fields,
+      networkClientOperations,
+      autoManagedNetworkClientRegistry,
     });
 
-    if (!existingNetworkConfiguration) {
-      this.#trackMetaMetricsEvent({
-        event: 'Custom Network Added',
-        category: 'Network',
-        referrer: {
-          url: referrer,
-        },
-        properties: {
-          // TODO: Either fix this lint violation or explain why it's necessary to ignore.
-          // eslint-disable-next-line @typescript-eslint/naming-convention
-          chain_id: chainId,
-          symbol: ticker,
-          source,
-        },
+    const replacementSelectedRpcEndpointWithIndex = networkClientOperations
+      .map(
+        (networkClientOperation, index) =>
+          [networkClientOperation, index] as const,
+      )
+      .find(([networkClientOperation, _index]) => {
+        return (
+          networkClientOperation.type === 'replace' &&
+          networkClientOperation.oldRpcEndpoint.networkClientId ===
+            this.state.selectedNetworkClientId
+        );
       });
+    const correctedReplacementSelectedRpcEndpointIndex =
+      replacementSelectedRpcEndpointIndex ??
+      replacementSelectedRpcEndpointWithIndex?.[1];
+
+    let rpcEndpointToSelect: RpcEndpoint | undefined;
+    if (correctedReplacementSelectedRpcEndpointIndex !== undefined) {
+      rpcEndpointToSelect =
+        updatedNetworkConfiguration.rpcEndpoints[
+          correctedReplacementSelectedRpcEndpointIndex
+        ];
+
+      // TODO: Test
+      if (rpcEndpointToSelect === undefined) {
+        throw new Error(
+          `Could not update network: \`replacementSelectedRpcEndpointIndex\` ${correctedReplacementSelectedRpcEndpointIndex} does not refer to an entry in \`rpcEndpoints\``,
+        );
+      }
     }
 
-    if (setActive) {
-      await this.setActiveNetwork(upsertedNetworkConfigurationId);
+    // TODO: Test
+    if (
+      rpcEndpointToSelect &&
+      rpcEndpointToSelect.networkClientId !== this.state.selectedNetworkClientId
+    ) {
+      await this.setActiveNetwork(rpcEndpointToSelect.networkClientId, {
+        updateState: (state) => {
+          this.#updateNetworkConfigurations({
```

Comment:
This was a fun one to figure out. I want to ensure that state updates get combined so that we don't have a brief moment where we've changed `selectedNetworkClientId` but `networkConfigurationsByChainId` is now out of date. I solved this by passing a function that takes state and changes the `networkConfigurationsByChainId` property, and then I call at the point where I set `selectedNetworkClientId`. But it'd be cool if we had a better way of doing this.